### ISSUE TITLE
Windows Worker Registration Failure

### ISSUE THREAD

**Issue by goffinf** [rancher/rancher#26739]
**What kind of request is this (question/bug/enhancement/feature request):**
Bug

**Steps to reproduce (least amount of steps as possible):**
Run the registration command provided from the Rancher UI for a WIndows worker node

`PowerShell -NoLogo -NonInteractive -Command "& {docker run -v c:/:c:/host -e AGENT_IMAGE=818213630802.dkr.ecr.eu-west-1.amazonaws.com/rancher/rancher-agent:v2.3.5 818213630802.dkr.ecr.eu-west-1.amazonaws.com/rancher/rancher-agent:v2.3.5 bootstrap --server https://pt02.containers.aws-euw1-np.avivacloud.com --token rk8gdlq7vx8txwkvmgf98t6t9c4d7t2dbfkzbjmskbngm45bscwtqj --worker | iex}"`

**Result:**
In the UI the Node is shown as registered. However the cattle-node-agent pad fails (as does the exportner-node-windows-cluster-monitoring workload).

When RDP'd onto the Windows worker, the connection continuously drops then reconnects until the registered node is deleted.

**Other details that may be helpful:**
The flannel configuration for the cluster is set as follows (from UI, Edit Cluster -> Edit as YAML)

```
  network:
    mtu: 0
    options:
      flannel_backend_port: '4789'
      flannel_backend_type: vxlan
      flannel_backend_vni: '4096'
    plugin: flannel
```
On the Windows worker node ...

**C:\etc\cni\net.d\10-flannel.conflist**

`{"cniVersion":"0.2.0","name":"vxlan0","plugins":[{"capabilities":{"dns":true},"delegate":{"dns":{"nameservers":["10.43.0.10"],"search":["svc.cluster.local"]},"policies":[{"value":{"ExceptionList":["10.42.0.0/16","10.43.0.0/16"],"Type":"OutBoundNAT"},"name":"EndpointPolicy"},{"value":{"DestinationPrefix":"10.43.0.0/16","NeedEncap":true,"Type":"ROUTE"},"name":"EndpointPolicy"}],"type":"win-overlay"},"type":"flannel"}]}`

**C:\etc\kube-flannel\net-conf.json**

`{"Network":"10.42.0.0/16","Backend":{"Name":"vxlan0","VNI":4096,"Port":4789,"Type":"vxlan"}}`

**C:\etc\rancher\wins\config**

`{"whiteList":{"processPaths":["c:\\etc\\wmi-exporter\\wmi-exporter.exe","c:\\etc\\kubernetes\\bin\\kube-proxy.exe","c:\\etc\\kubernetes\\bin\\kubelet.exe","c:\\etc\\nginx\\nginx.exe","c:\\opt\\bin\\flanneld.exe"]}}`

**C:\run\flannel\subnet.env**

**Does NOT EXIST**

The following pods continuously restart (nginx-proxy and the rancher-agent:v2.3.5 containers are stable):

- service-sidekick
- kubelet
- kube-proxy

**Logs**

**service-sidekick**

```
INFO: Start flanneld with: --kubeconfig-file=c:\etc\kubernetes\ssl\kubecfg-kube-node.yaml --ip-masq --kube-subnet-mgr --iptables-forward-rules=false --iface=10.63.131.11
I0421 11:57:30.902052   12744 main.go:454] Searching for interface using 10.63.131.11
I0421 11:57:30.914746   12744 main.go:531] Using interface with name Ethernet 3 and address 10.63.131.11
I0421 11:57:30.914746   12744 main.go:548] Defaulting external address to interface address (10.63.131.11)
I0421 11:57:30.937541   12744 kube.go:119] Waiting 10m0s for node controller to sync
I0421 11:57:30.937541   12744 kube.go:306] Starting kube subnet manager
I0421 11:57:31.943455   12744 kube.go:126] Node controller sync successful
I0421 11:57:31.943455   12744 main.go:246] Created subnet manager: Kubernetes Subnet Manager - ec2amaz-a8qr2l2
I0421 11:57:31.943455   12744 main.go:249] Installing signal handlers
I0421 11:57:31.943455   12744 main.go:390] Found network config - Backend type: vxlan
I0421 11:57:31.943637   12744 vxlan_windows.go:121] VXLAN config: Name=vxlan0 VNI=4096 Port=4789 GBP=false DirectRouting=false
E0421 11:57:31.956774   12744 main.go:291] Error registering network: failed to acquire lease: node "ec2amaz-a8qr2l2" not found
I0421 11:57:31.956774   12744 main.go:370] Stopping shutdownHandler...
...
```

**kube-proxy**

```
INFO: Waiting for servcie-sidekick to prepare the entrypoint.ps1 ...
FATA: Could not get vxlan0 HNS network: FATA[2020-04-21T11:45:43+01:00] rpc error: code = InvalidArgument desc = could not get HNS network vxlan0 via v2 api: Network Id  not found
INFO: Waiting for servcie-sidekick to prepare the entrypoint.ps1 ...
...
```

**kubelet**

```
E0421 11:46:56.130403   12392 cni_windows.go:59] error while adding to cni network: open /run/flannel/subnet.env: The system cannot find the path specified.
W0421 11:46:56.130403   12392 docker_sandbox.go:394] failed to read pod IP from plugin/docker: networkPlugin cni failed on the status hook for pod "exporter-node-windows-cluster-monitoring-kd9j9_cattle-prometheus": open /run/flannel/subnet.env: The system cannot find the path specified.
I0421 11:46:56.136273   12392 kubelet.go:1951] SyncLoop (PLEG): "exporter-node-windows-cluster-monitoring-kd9j9_cattle-prometheus(3050aab5-0978-476c-be73-4db06425392c)", event: &pleg.PodLifecycleEvent{ID:"3050aab5-0978-476c-be73-4db06425392c", Type:"ContainerDied", Data:"b20409a54f8fdd7ddd024c8d6eb0aad83c529eb1cc5cb7b3d02975cceb8862aa"}
W0421 11:46:56.137243   12392 pod_container_deletor.go:75] Container "b20409a54f8fdd7ddd024c8d6eb0aad83c529eb1cc5cb7b3d02975cceb8862aa" not found in pod's containers
I0421 11:46:56.137243   12392 kuberuntime_manager.go:442] No ready sandbox for pod "exporter-node-windows-cluster-monitoring-kd9j9_cattle-prometheus(3050aab5-0978-476c-be73-4db06425392c)" can be found. Need to start a new one
E0421 11:46:56.259451   12392 reflector.go:153] object-"cattle-logging"/"rancher-logging-fluentd-ssl": Failed to list *v1.Secret: Get https://127.0.0.1:6443/api/v1/namespaces/cattle-logging/secrets?fieldSelector=metadata.name%3Drancher-logging-fluentd-ssl&limit=500&resourceVersion=0: read tcp 127.0.0.1:64595->127.0.0.1:6443: wsarecv: An established connection was aborted by the software in your host machine.
E0421 11:46:56.434034   12392 remote_runtime.go:105] RunPodSandbox from runtime service failed: rpc error: code = Unknown desc = failed to set up sandbox container "6c111cc1b3ff59838059a4941ea6a0b8a5a850a8ceb8b9d0c76314588c5b768d" network for pod "rancher-logging-fluentd-windows-fj7bt": networkPlugin cni failed to set up pod "rancher-logging-fluentd-windows-fj7bt_cattle-logging" network: open /run/flannel/subnet.env: The system cannot find the path specified.
E0421 11:46:56.434034   12392 kuberuntime_sandbox.go:68] CreatePodSandbox for pod "rancher-logging-fluentd-windows-fj7bt_cattle-logging(64c723ad-351f-4ad6-b4b9-1b0c97511cce)" failed: rpc error: code = Unknown desc = failed to set up sandbox container "6c111cc1b3ff59838059a4941ea6a0b8a5a850a8ceb8b9d0c76314588c5b768d" network for pod "rancher-logging-fluentd-windows-fj7bt": networkPlugin cni failed to set up pod "rancher-logging-fluentd-windows-fj7bt_cattle-logging" network: open /run/flannel/subnet.env: The system cannot find the path specified.
E0421 11:46:56.434034   12392 kuberuntime_manager.go:729] createPodSandbox for pod "rancher-logging-fluentd-windows-fj7bt_cattle-logging(64c723ad-351f-4ad6-b4b9-1b0c97511cce)" failed: rpc error: code = Unknown desc = failed to set up sandbox container "6c111cc1b3ff59838059a4941ea6a0b8a5a850a8ceb8b9d0c76314588c5b768d" network for pod "rancher-logging-fluentd-windows-fj7bt": networkPlugin cni failed to set up pod "rancher-logging-fluentd-windows-fj7bt_cattle-logging" network: open /run/flannel/subnet.env: The system cannot find the path specified.
E0421 11:46:56.434034   12392 pod_workers.go:191] Error syncing pod 64c723ad-351f-4ad6-b4b9-1b0c97511cce ("rancher-logging-fluentd-windows-fj7bt_cattle-logging(64c723ad-351f-4ad6-b4b9-1b0c97511cce)"), skipping: failed to "CreatePodSandbox" for "rancher-logging-fluentd-windows-fj7bt_cattle-logging(64c723ad-351f-4ad6-b4b9-1b0c97511cce)" with CreatePodSandboxError: "CreatePodSandbox for pod \"rancher-logging-fluentd-windows-fj7bt_cattle-logging(64c723ad-351f-4ad6-b4b9-1b0c97511cce)\" failed: rpc error: code = Unknown desc = failed to set up sandbox container \"6c111cc1b3ff59838059a4941ea6a0b8a5a850a8ceb8b9d0c76314588c5b768d\" network for pod \"rancher-logging-fluentd-windows-fj7bt\": networkPlugin cni failed to set up pod \"rancher-logging-fluentd-windows-fj7bt_cattle-logging\" network: open /run/flannel/subnet.env: The system cannot find the path specified."
E0421 11:46:56.459546   12392 reflector.go:153] object-"cattle-logging"/"rancher-logging-fluentd": Failed to list *v1.Secret: Get https://127.0.0.1:6443/api/v1/namespaces/cattle-logging/secrets?fieldSelector=metadata.name%3Drancher-logging-fluentd&limit=500&resourceVersion=0: read tcp 127.0.0.1:64596->127.0.0.1:6443: wsarecv: An existing connection was forcibly closed by the remote host.
E0421 11:46:56.658954   12392 reflector.go:153] object-"cattle-prometheus"/"exporter-node-windows-cluster-monitoring": Failed to list *v1.ConfigMap: Get https://127.0.0.1:6443/api/v1/namespaces/cattle-prometheus/configmaps?fieldSelector=metadata.name%3Dexporter-node-windows-cluster-monitoring&limit=500&resourceVersion=0: read tcp 127.0.0.1:64597->127.0.0.1:6443: wsarecv: An existing connection was forcibly closed by the remote host.
FATA[2020-04-21T11:46:59+01:00] rpc error: code = Unavailable desc = transport is closing
...
NFO: Args: --client-ca-file=c:/etc/kubernetes/ssl/kube-ca.pem --resolv-conf= --feature-gates=SupportPodPidsLimit=true,RotateKubeletServerCertificate=true --authorization-mode=Webhook --image-pull-progress-deadline=30m --pod-max-pids=2048 --cluster-domain=cluster.local --system-reserved=cpu=1000m,memory=2Gi,ephemeral-storage=2Gi --make-iptables-util-chains=true --v=2 --kube-reserved=cpu=500m,memory=500Mi,ephemeral-storage=1Gi --protect-kernel-defaults=true --fail-swap-on=false --cert-dir=c:/var/lib/kubelet/pki --cluster-dns=10.43.0.10 --tls-cipher-suites=TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_RSA_WITH_AES_256_GCM_SHA384,TLS_RSA_WITH_AES_128_GCM_SHA256 --cni-conf-dir=c:/etc/cni/net.d --event-qps=0 --hostname-override=ec2amaz-a8qr2l2 --kubeconfig=c:/etc/kubernetes/ssl/kubecfg-kube-node.yaml --enforce-node-allocatable= --streaming-connection-idle-timeout=30m --address=0.0.0.0 --cloud-provider= --pod-infra-container-image=818213630802.dkr.ecr.eu-west-1.amazonaws.com/rancher/kubelet-pause:v0.1.3 --volume-plugin-dir=c:/var/lib/kubelet/volumeplugins --read-only-port=0 --network-plugin=cni --cgroups-per-qos=false --root-dir=c:/var/lib/kubelet --cni-bin-dir=c:/opt/cni/bin --anonymous-auth=false --authentication-token-webhook=true --node-ip=10.63.131.11
FATA[2020-04-21T11:48:31+01:00] rpc error: code = Internal desc = could not rename binary: could not backup the existing target file: rename c:\etc\kubernetes\bin\rancher-wins-kubelet.exe C:\Windows\TEMP\rancher-wins-kubelet.exe: Access is denied.
INFO: Waiting for servcie-sidekick to prepare the entrypoint.ps1 ...
```

**Environment information**
- Rancher version (`rancher/rancher`/`rancher/server` image tag or shown bottom left in the UI): **Rancher version: 2.3.5

Platform: **Amazon AWS**

OS: **Control Plane and Etcd + 1 worker - RHEL 7.7**
      **Windows workers - Windows Sever 2019 DataCenter - 1809 (Build 17763.973)**

- Installation option (single install/HA):

HA

**Cluster information**
- Cluster type (Hosted/Infrastructure Provider/Custom/Imported):  **Custom**
- Machine type (cloud/VM/metal) and specifications (CPU/memory): **All nodes AWS m52xLarge**
- Kubernetes version (use `kubectl version`): **1.17.2**

- Docker version (use `docker version`): **19.3.5**

**Comment by AMoghrabi** [comment]
Hey, did you end up finding a fix for this? Seeing something similar for flannel host-gw.

**Comment by goffinf** [comment]
Hey,

Yes, Windows node registration is now working successfully as part of instance launch provisioning. It took a couple of screen share sessions with the Rancher Engineering team in the US (and India) but we finally got there .. the org I work for has an Enterprise support contract with Rancher (which is why all the detail through that process is in the zendesk ticket rather than here).

Anyway, although there were a number of related issues, perhaps the most significant thing we missed was to ensure that we had ALL of the required security groups with the right protocol and applied to ALL node types. We had followed the docs fully (or so we thought) but with the one exception that we had missed in the additional port for VXLAN flannel (UDP 4789). Per Rancher advice we applied that to all  Control Plane, Etcd and Worker nodes and that did the trick. The /run/flannel/subnet.env was automatically created and populated with the correct values as were other related conf files.

We did go down a lot of other avenues, looking at MTU, corporate proxy (and no_proxy) settings, ... we also early on discovered that our Anti-virus/IDP software (Trend Micro) was blocking part of the registration process, so we had to create a policy to disable parts of that (we are still looking into that with Trend and discussing with our CISO). We also noted that registration messed with config for the AWS metadata service routes which then had the side effect of causing interactions with Amazon ECR to fail (we have had to add that back in post registration as part of our launch provisioning).

The 'other' related issues which we found included ...

1.  dnsPolicy issue for the cattle-cluster-agent

When set to ClusterFirst this workload would 'flap'. Setting it to 'Default' fixed that, however, one we had the registration working correctly AND did an upgrade to Rancher 2.4.3 / K8s 1.17.4, that went away and works just fine with ClusterFirst

2.  The Rancher UI would hang when we navigated to Registry Credentials.

That was resolved in Rancher v2.4.3

3.  Rancher System workloads 'flapping'

- exporter-node-windows-cluster-monitoring

Resolved when the registration was fixed.

- fluentd logging (cant remember the exact name cos we now have it turned off)

There is a known K8s issue for this one when using VXLAN which is apparently fixed in K8s 1.18.x. For now we have turned logging OFF (we did have it configured for Splunk). By the time we get closer to Prod for Windows clusters we will likely be running the Splunk K8s agent or another vendor stack here so its not a big concern for now.

4.  The metric api service discovery failed.

Resolved once registration was working properly.

5.  ECR authentication failed.

Resolved once we added the metadata service route back onto the host.  You can check with ..

route -4 print.

6.  MTU (disparity between Linux and Windows nodes).

This turned out to be a 'wild goose chase'. The values were correct (9001 and 1450)

HTHs

Fraser

**Comment by AMoghrabi** [comment]
Cheers
